### PR TITLE
fixed queue issues and did small refactor

### PR DESCRIPTION
**Pull request by igorJarek**

### CODE REVIEW

File: src/Queue.h [hunk 1]
```diff
@@ -128,58 +128,48 @@ template<size_t PARAMETER_SIZE>
...
-                  << std::endl;
-        m_mutex.unlock();
-    } else {
-        m_queue.push(request);
-        m_mutex.unlock();
-        m_cv.notify_one();
+        m_queue.emplace(sender_pid, data, length);
     }
+
+    m_condition_variable.notify_one();
```
Comment (rphh) [inline src/Queue.h:158]: I think that manual mutex lock/unlock was used for a reason. The mutex shall be unlocked before notify of condition variable.

File: src/Queue.h [hunk 2]
```diff
@@ -128,58 +128,48 @@ template<size_t PARAMETER_SIZE>
...
-    if(m_queue.size() >= m_max_elements) {
-        std::cerr << "Message loss in " << m_queue_name << " - queue is full, " << m_max_elements << " allowed"
-                  << std::endl;
-        m_mutex.unlock();
-    } else {
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        if(is_full()) {
+            return;
```
Comment (Lurkerpas) [inline src/Queue.h:135]: Why are we removing the message?

File: src/Request.h [hunk 3]
```diff
@@ -49,6 +50,13 @@ struct Request final
      */
     Request();
 
+    /**
+     * @brief Constructor
+     *
+     * Constructs Request with provided arguments
+     */
+    Request(const asn1SccPID sender_pid, const uint8_t* data, size_t length);
```
Comment (Lurkerpas) [inline src/Request.h:58]: length could be also const, there could probably be one more const in data (?)

File: src/Request.h [hunk 4]
```diff
@@ -159,6 +181,19 @@ Request<PARAMETER_SIZE>::set_sender_pid(asn1SccPID sender_pid)
 {
     m_sender_pid = sender_pid;
 }
+
+template<size_t PARAMETER_SIZE>
+void
+Request<PARAMETER_SIZE>::check_length(size_t length) const
+{
+    if(length > PARAMETER_SIZE) {
```
Comment (Lurkerpas) [inline src/Request.h:189]: Why not assert?

File: src/StartBarrier.cc [hunk 5]
```diff
@@ -21,19 +21,30 @@
...
 
 namespace taste {
 void
 StartBarrier::initialize(size_t number, InitCallback init_callback)
 {
     m_init_callback = init_callback;
-    pthread_barrier_init(&m_init_barrier, nullptr, number);
+
+    const int error_code = pthread_barrier_init(&m_init_barrier, nullptr, number);
+    if(error_code != 0) {
```
Comment (Lurkerpas) [inline src/StartBarrier.cc:33]: Why not assert?

File: src/StartBarrier.cc [hunk 6]
```diff
@@ -21,19 +21,30 @@
...
+    }
 }
 
 void
 StartBarrier::wait()
 {
-    pthread_barrier_wait(&m_init_barrier);
+    const int error_code = pthread_barrier_wait(&m_init_barrier);
+    if(error_code != PTHREAD_BARRIER_SERIAL_THREAD && error_code != 0) {
+        std::cerr << "Barrier Wait has been failed. Error code : " << error_code << std::endl;
```
Comment (Lurkerpas) [inline src/StartBarrier.cc:44]: Why not assert?

File: src/Queue.h [hunk 7]
```diff
@@ -192,6 +182,21 @@ Queue<PARAMETER_SIZE>::is_empty() const
     std::unique_lock<std::mutex> lock(m_mutex);
     return m_queue.empty();
 }
+
+template<size_t PARAMETER_SIZE>
+bool
+Queue<PARAMETER_SIZE>::is_full() const
+{
+    if(m_queue.size() >= m_max_elements) {
```
[the remark this answers is not in the document]
Comment (igorJarek) [reply src/Queue.h:190]: Renamed for check_for_message_loss()

File: src/Request.h (hunk 3 above)
Comment (igorJarek) [reply src/Request.h]: Done

File: src/Request.h (hunk 4 above)
Comment (igorJarek) [reply src/Request.h:189]: To be consistent with code in Thread.cc

File: src/StartBarrier.cc (hunk 5 above)
Comment (igorJarek) [reply src/StartBarrier.cc:33]: To be consistent with code in Thread.cc

File: src/StartBarrier.cc (hunk 6 above)
Comment (igorJarek) [reply src/StartBarrier.cc:44]: To be consistent with code in Thread.cc

File: src/Queue.h (hunk 1 above)
Comment (igorJarek) [reply src/Queue.h:158]: You are right, and this behavior is preserved.

**Review by Lurkerpas — APPROVED** [review]

File: src/Queue.h (hunk 1 above)
Comment (rphh) [reply src/Queue.h:158]: Ok, separate block for mutex.

**Review by rphh — APPROVED** [review]

**Pull request merged**